Source/core/Sync.cpp

```diff
@@ -782,6 +784,224 @@ namespace Core {
         return (nResult);
     }
 
+    //----------------------------------------------------------------------------
+    //----------------------------------------------------------------------------
+    // SharedSemaphore class
+    //----------------------------------------------------------------------------
+    //----------------------------------------------------------------------------
+
+    //----------------------------------------------------------------------------
+    // CONSTRUCTOR & DESTRUCTOR
+    //----------------------------------------------------------------------------
+
+
+    SharedSemaphore::SharedSemaphore(const TCHAR sourceName[], const uint32_t initCount, VARIABLE_IS_NOT_USED const uint32_t maxCount)
+    {
+        ASSERT(initCount <= 1);
+        ASSERT(maxCount == 1);
+
+#ifdef __WINDOWS__
+        _semaphore = (::CreateSemaphore(nullptr, initCount, maxCount, sourceName));
+        ASSERT(_semaphore != nullptr);
+#else
+        _name = "/" + string(sourceName);
+        _semaphore = sem_open(_name.c_str(), O_CREAT | O_RDWR, 0644,  initCount);
+        ASSERT(_semaphore != SEM_FAILED);
+#endif
+    }
+
+    SharedSemaphore::SharedSemaphore(const TCHAR sourceName[])
+    {
+#ifdef __WINDOWS__
+        _semaphore = ::OpenSemaphore(SEMAPHORE_ALL_ACCESS, FALSE, sourceName);
+        ASSERT(_semaphore != nullptr);
+#else
+        _semaphore = sem_open(sourceName, 0);
+        ASSERT(_semaphore != nullptr);
+#endif
+    }
+
+#ifndef __WINDOWS__
+    SharedSemaphore::SharedSemaphore(void* storage, const uint32_t initCount, const uint32_t maxCount)
+        : _semaphore(storage), _name("")
+    {
+        ASSERT(storage != nullptr);
+        ASSERT(initCount <= 1);
+        ASSERT(maxCount == 1);
+
+        if (initCount != 0 && maxCount != 0) {
```

Comment:
This the culprit: the semaphore still needs initialization even if initCount is 0! (so only check for maxCount != 0)